snuba/lw_deletions/batching.py

```diff
@@ -0,0 +1,157 @@
+from __future__ import annotations
+
+import time
+from typing import Callable, Generic, MutableSequence, Optional, TypeVar, Union
+
+from arroyo.processing.strategies.abstract import ProcessingStrategy
+from arroyo.processing.strategies.buffer import Buffer
+from arroyo.types import BaseValue, FilteredPayload, Message, TStrategyPayload
+
+ValuesBatch = MutableSequence[BaseValue[TStrategyPayload]]
+
+
+TPayload = TypeVar("TPayload")
+TResult = TypeVar("TResult")
+
+
+Accumulator = Callable[[TResult, BaseValue[TPayload]], TResult]
+
+
+class ReduceRowsBuffer(Generic[TPayload, TResult]):
+    def __init__(
+        self,
+        accumulator: Accumulator[TResult, TPayload],
+        initial_value: Callable[[], TResult],
+        max_batch_size: int,
+        max_batch_time: float,
+        increment_by: Optional[Callable[[BaseValue[TPayload]], int]] = None,
+    ):
+        self.accumulator = accumulator
+        self.initial_value = initial_value
+        self.max_batch_size = max_batch_size
+        self.max_batch_time = max_batch_time
+        self.increment_by = increment_by
+
+        self._buffer = initial_value()
+        self._buffer_size = 0
+        self._buffer_until = time.time() + max_batch_time
```

Comment:
This is copied from the arroyo code - and ultimately will be deleting this file entirely so Im just going to keep as is since it's consistent with arroyo